### PR TITLE
Fix SiLU gradient underflow for large negative inputs

### LINKED ISSUE #124835: tf.nn.silu returns wrong gradient for large negative float64 input
### Issue type

Bug

### Have you reproduced the bug with TensorFlow Nightly?

Yes

### Source

source

### TensorFlow version

tf 2.21.0

### Custom code

Yes

### OS platform and distribution

Linux Ubuntu 22.04

### Mobile device

_No response_

### Python version

Python 3.13

### Bazel version

_No response_

### GCC/compiler version

_No response_

### CUDA/cuDNN version

_No response_

### GPU model and memory

_No response_

### Current behavior?

`tf.nn.silu` returns an incorrect gradient for a large negative float64 input.

For SiLU / swish, the function is `x * sigmoid(x)`. At `x = -709.0`, the analytic high-precision reference for the derivative is approximately `-8.614807714413836e-306`, which is a finite nonzero float64 value.

However, TensorFlow autodiff returns `-0.0` for this gradient. PyTorch computes the expected finite value for the same derivative, so this appears to be a backward-pass issue rather than a float64 limitation.

Expected behavior?

The gradient of `tf.nn.silu` at `x = -709.0` should be close to the analytic reference value `-8.614807714413836e-306`, not `-0.0`.

### Standalone code to reproduce the issue

```shell
import os
os.environ.setdefault("CUDA_VISIBLE_DEVICES", "-1")
os.environ.setdefault("TF_CPP_MIN_LOG_LEVEL", "3")

import tensorflow as tf

x = -709.0
truth = -8.614807714413836e-306
rtol = 1e-6

v = tf.Variable(x, dtype=tf.float64)

with tf.GradientTape() as tape:
    y = tf.nn.silu(v)

g = tape.gradient(y, v)

actual = float(g)
rel_err = abs(actual - truth) / abs(truth)

print(f"x: {x!r}")
print(f"tensorflow gradient: {actual!r}")
print(f"expected gradient: {truth!r}")
print(f"relative error: {rel_err:.3e}")

if rel_err > rtol:
    print("BUG REPRODUCED: tf.nn.silu gradient is wrong for large negative float64 input")
else:
    print("not reproduced")
```

### Relevant log output

```shell
x: -709.0
tensorflow gradient: -0.0
expected gradient: -8.614807714413836e-306
relative error: 1.000e+00
BUG REPRODUCED: tf.nn.silu gradient is wrong for large negative float64 input
```

### PR DESCRIPTION
**Pull request by GodlyDonuts**
Fixes #124835.

## What changed

- Keep the existing fast SiLU path when `sigmoid(beta * x)` is representable.
- When the sigmoid intermediate underflows but its scaled product is still representable, evaluate that product in the log domain.
- Apply the guarded fallback to the forward value, feature gradient, and beta gradient.
- Add float32 and float64 regression coverage in graph and eager modes.

## Validation

- Reproduced the reported TensorFlow 2.20.0 result at float64 `x = -709`: the feature gradient was `-0.0`.
- Verified the patched implementation returns a finite feature gradient of `-8.614807714413413e-306` (analytic reference: `-8.614807714413836e-306`).
- Verified float32 `x = -90`, the forward result, beta gradient, vector inputs, custom beta values, infinities, signed zero, and NaN behavior.
- Passed the new regression test in graph and eager modes.
- Passed the existing `SwishGradOpTest.testSwishGrad` and full `SwishTest` suite.
- `python3 -m py_compile tensorflow/python/ops/nn_impl.py tensorflow/python/ops/nn_grad_test.py`
- TensorFlow PyLint configuration: 10.00/10.
- `git diff --check`

### CODE REVIEW

File: tensorflow/python/ops/nn_impl.py [hunk 1]
```diff
@@ -459,6 +459,23 @@ def swish(features, beta=1.0):
...
+    """Computes `multiplier * sigmoid(logits)` without early underflow."""
+    # sigmoid(logits) may underflow even when the scaled result is representable.
+    # In that range sigmoid(logits) and exp(logits) differ by less than the
+    # dtype's precision, so evaluate the product in the log domain.
+    sigmoid_underflow = math_ops.logical_and(
+        math_ops.equal(sigmoid_logits, 0.0), math_ops.is_finite(logits))
+    safe_multiplier = array_ops.where_v2(
+        sigmoid_underflow, multiplier, array_ops.ones_like(multiplier))
+    safe_logits = array_ops.where_v2(
+        sigmoid_underflow, logits, array_ops.zeros_like(logits))
```
[the remark this answers is not in the document]
Comment (GodlyDonuts) [reply tensorflow/python/ops/nn_impl.py]: Addressed in 4733671d99f. The fallback now broadcasts dtype-matched scalar one/zero values instead of materializing full-size tensors. The underflow regression, existing Swish gradient test, and full Swish test suite pass in graph and eager modes.

**Comment by GodlyDonuts** [comment]
@dmiltr3, would you be able to review this change when you have a chance? Thank you.

**Review by dmiltr3 — APPROVED** [review]
LGTM. Thanks!
